xarray/core/dataset.py

```diff
@@ -728,20 +723,26 @@ def _replace_vars_and_dims(self, variables, coord_names=None, dims=None,
                 self._dims = dims
             if attrs is not self.__default:
                 self._attrs = attrs
-            if encoding is not None:
+            if indexes is not self.__default:
+                self._indexes = indexes
+            if encoding is not self.__default:
                 self._encoding = encoding
-            self._indexes = indexes
             obj = self
         else:
             if variables is None:
                 variables = self._variables.copy()
             if coord_names is None:
                 coord_names = self._coord_names.copy()
-            if attrs is self.__default_attrs:
-                attrs = self._attrs_copy()
-            obj = self._construct_direct(variables, coord_names=coord_names,
-                                         dims=dims, attrs=attrs,
-                                         indexes=indexes, encoding=encoding)
+            if dims is None:
+                dims = self._dims.copy()
+            if attrs is self.__default:
+                attrs = copy.copy(self._attrs)
+            if indexes is self.__default:
+                indexes = copy.copy(self._indexes)
+            if encoding is self.__default:
+                encoding = copy.copy(self._encoding)
+            obj = self._construct_direct(
+                variables, coord_names, dims, attrs, indexes, encoding)
```

Comment:
@shoyer - just to close the loop on this. Your indexes refactor added the `encoding` argument here. You'll see in this merge diff that I had also added this argument in the same place.